src/plugins/ucc_discovery_plugin.cpp

```diff
@@ -106,38 +112,62 @@ void UccDiscoveryPlugin::_validate_ucc_candidates(const UccCandidates& ucc_candi
 
     const auto& soft_key_constraints = table->soft_key_constraints();
 
-    // Skip already discovered UCCs.
-    if (std::any_of(soft_key_constraints.cbegin(), soft_key_constraints.cend(),
-                    [&column_id](const auto& key_constraint) {
-                      const auto& columns = key_constraint.columns();
-                      return columns.size() == 1 && *columns.cbegin() == column_id;
-                    })) {
-      message << " [skipped (already known) in " << candidate_timer.lap_formatted() << "]";
+    // Find UCC matching the candidate, if it already exists on the table.
+    const auto& existing_ucc = std::find_if(soft_key_constraints.cbegin(), soft_key_constraints.cend(),
+                                            [&column_id](const auto& key_constraint) {
+                                              const auto& columns = key_constraint.columns();
+
+                                              return columns.size() == 1 && *columns.cbegin() == column_id;
+                                            });
+
+    // Check if MVCC data tells us that the existing UCC is guaranteed to be still valid.
+    // If it is, we can skip the expensive revalidation of the UCC.
+    if (existing_ucc != soft_key_constraints.cend() && table->constraint_guaranteed_to_be_valid(*existing_ucc)) {
+      message << " [skipped (already known and guaranteed to be still valid) in " << candidate_timer.lap_formatted()
+              << "]";
       Hyrise::get().log_manager.add_message("UccDiscoveryPlugin", message.str(), LogLevel::Info);
+      existing_ucc->revalidated_on(transaction_context->snapshot_commit_id());
       continue;
     }
 
+    // If no UCC already exists or the existing UCC is not guaranteed to be still valid, we have to now (re-)validate
+    // the UCC candidate.
     resolve_data_type(table->column_data_type(column_id), [&](const auto data_type_t) {
       using ColumnDataType = typename decltype(data_type_t)::type;
 
+      bool ucc_is_currently_valid;
       // Utilize efficient check for uniqueness inside each dictionary segment for a potential early out.
+      // If that does not allow us to reject the UCC right away, we have to run the more expensive
+      // cross-segment duplicate check (next clause of the if-condition).
```

Comment:
Quick question regarding what is meant by `(next clause of the if-condition).`
Is it the next else-if clause or the next if block? Because if it's the latter I'd propose to keep this comment, that has been deleted `// If we reach here, we have to run the more expensive cross-segment duplicate check.`

Reply:
The next else-if clause is meant. In the first clause of the if-condition, we try to cast each segment to a dictionary segment and check if the size of an unmutable dictionary segment is unequal to the unique values of the segment. If we haven't found a unmutable segment that fulfills this condition, there could still be mutable segments, that contain duplicates which we have to check in the next else-if clause.